flax/nn/base.py

```diff
@@ -943,7 +945,13 @@ def store(self, value):
     frame = _top_frame('store')
     if not self._mutable:
       raise ValueError('Collection is not mutable. Use the `mutate` method to'
-                       'create a mutable copy.')
+                       ' create a mutable copy.')
+    master = utils._tracer_of_value(value)
+    value_level = master.level if master else -1000
+    state_level = self._master.level if self._master else -1000
+    if value_level > state_level:
+      raise ValueError('Stateful operations are not allowed when the Collection'
+                       ' is created outside of the current Jax transformation')
```

Comment:
It would be great if this error message guided users to a doc link that shows one or two examples of "how to interleave JAX transformations with Flax modules".

We don't need to do that in this pull request, but if not, I think we should track this in a new issue.

Reply:
The doc is in the making. But I staged it out to a seperate PR indeed